src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -344,133 +344,312 @@ public class SqlMutationEngine : IMutationEngine
+                                selectOperationResponse = await _queryEngine.ExecuteFollowUpReadAsync(findRequestContext);
+                            }
+
+                            transactionScope.Complete();
+                        }
                     }
 
-                    if (mutationResultRow.Columns.Count == 0)
+                    // All the exceptions that can be thrown by .Complete() and .Dispose() methods of transactionScope
+                    // derive from TransactionException. Hence, TransactionException acts as a catch-all.
+                    // When an exception related to Transactions is encountered, the mutation is deemed unsuccessful and
+                    // a DataApiBuilderException is thrown
+                    catch (TransactionException)
                     {
-                        throw new DataApiBuilderException(
-                            message: "Could not insert row with given values.",
-                            statusCode: HttpStatusCode.Forbidden,
-                            subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure
-                            );
+                        throw _dabExceptionWithTransactionErrorMessage;
                     }
 
-                    string primaryKeyRoute = ConstructPrimaryKeyRoute(context, mutationResultRow.Columns);
-                    // location will be updated in rest controller where httpcontext is available
-                    return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(mutationResultRow.Columns).Value);
-                }
+                    string primaryKeyRoute = ConstructPrimaryKeyRoute(context, mutationResultRow!.Columns);
 
-                if (context.OperationType is EntityActionOperation.Update || context.OperationType is EntityActionOperation.UpdateIncremental)
-                {
-                    // Nothing to update means we throw Exception
-                    if (mutationResultRow is null || mutationResultRow.Columns.Count == 0)
+                    // When read permission is configured without a database policy, a subsequent select query will not be executed.
+                    // So, if the read action has include/exclude fields configured, additional fields present in the response
+                    // need to be removed.
+                    if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                    {
+                        IEnumerable<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read);
+                        foreach (string columnInResponse in mutationResultRow!.Columns.Keys)
+                        {
+                            if (!allowedExposedColumns.Contains(columnInResponse))
+                            {
```

Comment:
How will we return the LocationHeader if `allowedExposedColumns` doesn't contain the primary key columns?

Reply:
For PUT and PATCH API requests, the users are aware of the Pks as it is required to be passed in the request URL. In case of tables with auto-gen PKs, PUT or PATCH will not result in an insert but error out. Seeing that Location Header does not provide users with any additional information, it is set as an empty string always.

For POST API requests, the primary key route calculated will be an empty string only when the read action for the role does not have access to one or more PKs. So, in this case too, the Location Header is set to an empty string.